python/pfs/utils/optimizeBlackDots.py

```diff
@@ -0,0 +1,501 @@
+
+            if np.sum(isGood_init) < min_n_points:
+                mcs_data_extended[0, i][isGood_init] = mcs_data_all[0, fib][isGood_init]
+                mcs_data_extended[1, i][isGood_init] = mcs_data_all[1, fib][isGood_init]
+                predicted_position_x = np.full(n_obs, np.nan)
+                predicted_position_y = np.full(n_obs, np.nan)
+                mcs_data_extended[0, i][~isGood_init] = predicted_position_x[~isGood_init]
+                mcs_data_extended[1, i][~isGood_init] = predicted_position_y[~isGood_init]
+                mcs_data_extended[2, i] = isGood_init
+                continue
+
+            # Select the point that are more than outlier_distance (default is 2 mm)
+            # away from median of all of the points and replace them with np.nan
+            # We assume these points have been attributed to wrong cobras
+            xMedian = np.median(x_measurments[isGood_init])
+            yMedian = np.median(y_measurments[isGood_init])
+            large_outliers_selection =\
+                (np.abs(y_measurments[isGood_init] - yMedian) > outlier_distance) |\
+                (np.abs(x_measurments[isGood_init] - xMedian) > outlier_distance)
+
+            # Remove the large outliers from the selection of good points
+            isGood = np.copy(isGood_init)
+            isGood[isGood_init == 1] = isGood_init[isGood_init == 1]*~large_outliers_selection
+
+            # Replace all unsuccessful measurments wih nan
+            x_measurments[~isGood] = np.nan
+            y_measurments[~isGood] = np.nan
+
+            # Specify the number of breaks in the data
+            if np.sum(np.diff(isGood)) == 1:
+                number_of_breaks = 1
+            else:
+                number_of_breaks = 2
+
+            if type_of_run == 'theta':
+                poly_order_inter = order_theta_inter
+                poly_order_extra = order_theta_extra
+            if type_of_run == 'phi':
+                poly_order_inter = order_phi_inter
+                poly_order_extra = order_phi_extra
+            # Create prediction for where the points should be, when you do not see points
+            # If you do not see points from both sides of the black dots, do simpler extrapolation
+            # If you see points from both sides of the black dots, do more complex interpolation
+            if number_of_breaks == 1:
+                deg2_fit_to_single_point_y, residuals_y =\
+                    self.poly_fit(n_obs, y_measurments, isGood, poly_order_extra)
+                deg2_fit_to_single_point_x, residuals_x =\
+                    self.poly_fit(n_obs, x_measurments, isGood, poly_order_extra)
+            else:
+                deg2_fit_to_single_point_y, residuals_y =\
+                    self.poly_fit(n_obs, y_measurments, isGood, poly_order_inter)
+                deg2_fit_to_single_point_x, residuals_x =\
+                    self.poly_fit(n_obs, x_measurments, isGood, poly_order_inter)
```

Comment:
These blocks are exactly the same except for the last argument, so condense. In general (i.e., there are exceptions to the rule), code duplication is evil, because it's more code to maintain and higher probability that there will be bugs introduced.